### PR TITLE
implement excel importer

### PR DESCRIPTION
**Pull request by lsmith77**
also refactored AbstractImporter to be more generic

fixes https://github.com/FriendsOfSylius/SyliusImportExportPlugin/issues/15

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,10 +6,15 @@
     "require": {
         "php": "^7.1",
         "sylius/sylius": "^1.0",
-        "portphp/portphp": "^1.1",
-        "portphp/csv": "^1.0"
+        "portphp/portphp": "^1.2"
+    },
+    "suggest": {
```
Comment (lsmith77) [inline composer.json]: waiting on @ddeboer to tag new releases

File: src/Importer/PaymentMethodsImporter.php [hunk 2]
```diff
@@ -14,7 +15,22 @@
...
     /** @var array */
-    protected $headerKeys = ['Code', 'Gateway', 'Name', 'Instructions'];
+    private $headerKeys = ['Code', 'Gateway', 'Name', 'Instructions'];
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function assertKeys(Reader $reader)
+    {
+        if (!method_exists($reader, 'getColumnHeaders')) {
```
Comment (lsmith77) [inline src/Importer/PaymentMethodsImporter.php]: this is quite inelegant .. not sure if it makes asking @ddeboer for another interface here .. or to check for specific instances of the relevant reader classes .. I guess lets see how this evolves with other formats?

File: src/Resources/config/services.yml [hunk 3]
```diff
@@ -10,6 +10,11 @@ services:
         arguments:
             - 0
 
+    sylius.factory.excel_reader:
+        class: Port\Excel\ExcelReaderFactory
+        arguments:
+            - 1
```
Comment (lsmith77) [inline src/Resources/config/services.yml]: I guess such stuff should be configurable ..

File: src/DependencyInjection/SyliusImportExportExtension.php [hunk 4]
```diff
@@ -19,6 +19,18 @@ public function load(array $configs, ContainerBuilder $container)
         $configuration = new Configuration();
         $this->processConfiguration($configuration, $configs);
 
+        if (!class_exists('Port\Csv\CsvReaderFactory')) {
```
Comment (lsmith77) [inline src/DependencyInjection/SyliusImportExportExtension.php]: this also feels a bit iffy

**Comment by lsmith77** [comment]
@solutionDrive-Alt

**Comment by lsmith77** [comment]
I am not sure how the model the tests for excel .. since excel is just a matter of different config compared to csv .. I guess behat tests also test the configuration .. so it makes sense to duplicate them but not the spec tests?

File: src/DependencyInjection/SyliusImportExportExtension.php (hunk 4 above)
Comment (lsmith77) [reply src/DependencyInjection/SyliusImportExportExtension.php]: https://twitter.com/lsmith/status/927481599624609792

File: src/Importer/PaymentMethodsImporter.php (hunk 2 above)
Comment (solutionDrive-Alt) [reply src/Importer/PaymentMethodsImporter.php]: maybe we could have our own ReaderInterface which is implemented by small wrapper-classes (which wrap the actual reader) and delegate the call to the method defined in our own Interface to the corresponding Method of the wrapped Reader (e.g. PortCsv or some XmlReader). I hope my thoughts are understandable and I don't misunderstand your concerns.

File: src/Resources/config/services.yml (hunk 3 above)
Comment (solutionDrive-Alt) [reply src/Resources/config/services.yml]: yeah probably.

**Comment by solutionDrive-Alt** [comment]
good question, I think most of the phpspec tests would look almost the same, especially if the value returned by the reader is also an array. But on the other hand it wouldn't hurt that much and we would be on the save side and we don't know for sure that we will always use the two phpport-readers.

File: src/Importer/PaymentMethodsImporter.php (hunk 2 above)
Comment (lsmith77) [reply src/Importer/PaymentMethodsImporter.php]: yeah .. I was pondering the same as an option.

semi-related .. I was also pondering to move this code into the `AbstractImporter` ..

Comment (solutionDrive-Alt) [reply src/Importer/PaymentMethodsImporter.php]: yes, centralising would be a good idea, if it is possible.

File: src/Importer/AbstractImporter.php [hunk 5]
```diff
@@ -58,21 +59,35 @@ public function __construct(
...
     }
 
     /**
+     * @param Reader $reader
+     *
+     * @throws ImporterException
+     */
+    protected function assertKeys(Reader $reader)
+    {
+        if (!method_exists($reader, 'getColumnHeaders')) {
```
Comment (lsmith77) [inline src/Importer/AbstractImporter.php]: hopefully we can soon remove this if https://github.com/portphp/portphp/pull/64 gets merged

**Comment by lsmith77** [comment]
adde Behat tests .. but left out spec tests as I feel this is redundant at least for now

File: composer.json (hunk 1 above)
Comment (solutionDrive-Alt) [reply composer.json]: is this a reason for you to keep this PR open, or should we merge it?

**Comment by lsmith77** [comment]
we now have behat tests .. skipped phpspec .. partially because I wasn’t entirely sure how to set them up and also because I really feel this is redundant atm because in our codebase we are taking exactly the same code paths.

**Comment by lsmith77** [comment]
@solutionDrive-Alt I guess its your turn to merge if you approve :)

**Comment by solutionDrive-Alt** [comment]
true, just thought your comment above in composer.json could be a reason for you to keep it open.

**Pull request merged**